src/seadrive-sso/auto-logon-dialog.cpp

```diff
@@ -0,0 +1,242 @@
+#endif
+const char *const kSeafilePreconfigureGroupName = "preconfigure";
+const char *const kPreconfigureServerAddr = "PreconfigureServerAddr";
+}
+
+AutoLogonDialog::AutoLogonDialog(const QUrl& url, QWidget* parent)
+    : QDialog(parent), login_url_(url)
+{
+    setWindowTitle(
+        QString("%1 %2").arg(getBrand()).arg(tr("auto login")));
+    setWindowIcon(QIcon(":/images/seafile.png"));
+    QHBoxLayout* layout = new QHBoxLayout;
+
+    QWidget* lspacer = new QWidget;
+    lspacer->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Minimum);
+    layout->addWidget(lspacer);
+
+    QLabel* label = new QLabel(tr("auto logining, please wait a moment"));
+    layout->addWidget(label);
+
+    QWidget* rspacer = new QWidget;
+    rspacer->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Minimum);
+    layout->addWidget(rspacer);
+
+    setLayout(layout);
+
+
+    setStyleSheet(
+        "QDialog { min-width: 300px; max-wdith: 300px; min-height: 130px; "
+        "max-height: 130px; }");
+
+    QTimer::singleShot(100, this, SLOT(startAutoLogon()));
+}
+
+/**
+* Read the server address from windows regsitry. The value shoud be
+* preconfigured by AD.
+*
+*   key: HKEY_LOCAL_MACHINE\SOFTWARE\\SeaDrive\\ServerUrl
+*   value: like http://sharefile.seafile.com.cn
+**/
+QUrl AutoLogonDialog::readServerUrlFromRegistry()
+{
+    QUrl url;
+    RegElement reg(HKEY_LOCAL_MACHINE, "SOFTWARE\\SeaDrive", "ServerUrl", "");
+    if (!reg.exists()) {
+        qWarning("failed to read server url from registry");
+        return url;
+    }
+    reg.read();
+    url = reg.stringValue();
+    if (!url.isValid()) {
+        errorAndExit(
+            tr("server address format error: %1").arg(reg.stringValue()));
+    }
+    return url;
+}
+
+QUrl AutoLogonDialog::readServerUrlFromConfigFile(const QString& key)
+{
+    QUrl url;
+    QString configure_file = QDir::home().filePath(kSeafileConfigureFileName);
+    if (!QFileInfo(configure_file).exists()) {
+        qWarning("config file not exists");
+        return url;
+    }
+    QSettings setting(configure_file, QSettings::IniFormat);
+    setting.beginGroup(kSeafilePreconfigureGroupName);
+    QVariant value = setting.value(key);
+    setting.endGroup();
+    url = value.toString();
+    if (!url.isValid()) {
+        errorAndExit(
+            tr("server address format error: %1").arg(url.toString()));
+    }
+    return url;
+}
+
+QUrl AutoLogonDialog::askForServerUrl()
+{
+    QUrl url;
+    QString str;
+    bool ok;
+    while (true) {
+        str = QInputDialog::getText(this, tr("Please input server address"),
+                                    tr("address"),
+                                    QLineEdit::Normal, str, &ok)
+                  .trimmed();
+        if (!ok) {
+            qWarning("user cancelled the server url input dialog.");
+            errorAndExit();
+            return QUrl();
+        }
+        if (str.isEmpty()) {
+            warn(tr("address cannot empty"));
```

Comment:
`address cannot empty` => `Please enter the server address`

Reply:
ok